docs/codeql/codeql-language-guides/customizing-library-models-for-javascript.rst

```diff
@@ -0,0 +1,392 @@
+  });
+
+This source is already known by the CodeQL JS analysis, but we'll show how it could be added as an extension. This can be achieved with the following extension:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/javascript-all
+        extensible: sourceModel
+      data:
+        - [
+            "global",
+            "Member[addEventListener].Argument[1].Parameter[0].Member[data]",
+            "remote-flow",
+          ]
+
+To break this down:
+
+- Since we're adding a new taint source, we add a tuple to the **sourceModel** extension point.
+- The first column, **"global"**, begins the search at references to the global object (also known as **window**). This is a special JavaScript object that contains all global variables and methods.
+- **Member[addEventListener]** selects accesses to the **addEventListener** member.
+- **Argument[1]** selects the second argument of calls to that member (the argument containing the callback).
+- **Parameter[0]** selects the first parameter of the callback (the parameter named **event**).
+- **Member[data]** selects accesses to the **data** property of the event object.
+- Finally, the kind **remote-flow** indicates that this is considered a source of remote flow.
+
+Continued example: Restricting the event type
+---------------------------------------------
+
+The model above treats all events as sources of remote flow, not just **message** events.
+For example, it would also pick up this irrelevant source:
+
+.. code-block:: js
+
+  window.addEventListener("onclick", function (event) {
+    let data = event.data; // <-- 'event.data' became a spurious taint source
+  });
+
+
+We can refine the model by adding the **WithStringArgument** component to restrict the set of calls being considered:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/javascript-all
+        extensible: sourceModel
+      data:
+        - [
+            "global",
+            "Member[addEventListener].WithStringArgument[0=message].Argument[1].Parameter[0].Member[data]",
+            "remote-flow",
+          ]
+
+The **WithStringArgument[0=message]** component here selects the subset of calls to **addEventListener** where the first argument is a string literal with the value **"message"**.
+
+Example: Using types to add MySQL injection sinks
+-------------------------------------------------
+
+In this example, we'll show how to add the following SQL injection sink:
+
+.. code-block:: ts
+
+  import { Connection } from "mysql";
+
+  function submit(connection: Connection, q: string) {
+    connection.query(q); // <-- add 'q' as a SQL injection sink
+  }
+
+We can recognize this using the following extension:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/javascript-all
+        extensible: sinkModel
+      data:
+        - ["mysql.Connection", "Member[query].Argument[0]", "sql-injection"]
```

Comment:
Question: Is this equivalent?

```suggestion
        - ["mysql", "Member[Connection].Member[query].Argument[0]", "sql-injection"]
```

And if so, do you have a recommendation for which is better?

Reply:
Maybe this is different because `mysql.Connection` is a type, not a value?

Reply:
Indeed they are different. What you wrote would correspond to a static method on the `Connection` class.

Reply:
Oh....that's subtle. It makes sense, but it's something that I think many people (including me) would miss without explicit documentation.

Reply:
Pushed an explanation in [d62b944](https://github.com/github/codeql/pull/11615/commits/d62b944b930b1f9eda97821d341ad2ce7add9103)